### PR TITLE
[termops] do not strip casts from evars (fix #14972)

### LINKED ISSUE #294: Refine elpi primitive does not act like Coq's refine
Consider an elpi tactic `t` calling `refine ((_ : A) B)` on the Coq goal for `A` and `B` given as arguments.
```coq
Elpi Tactic t.
Elpi Accumulate lp:{{
  solve (goal _ _ _ _ [trm A, trm B] as G) Gs :-
    coq.say {{ (_ : lp:A ) lp:B }},
    refine {{ (_ : lp:B) lp:B }} G Gs.
}}.
Elpi Typecheck.
```

Say I want to use it to generalise a goal.
```coq
Goal 0 = 0.
Proof.
  refine ((_ : forall x : nat, x = 0) 0).
  Show.
```
```
1 subgoal
  
  ============================
  forall x : nat, x = 0
```

If we replace the `refine` call with the elpi tactic, we get the following goal:
```coq
  elpi t (forall x : nat, x = 0) (0).
  Show.
```
```
1 focused subgoal
(shelved: 1)
  
  ============================
  forall x : nat, x = x
```

It looks like the call to refine did something different. Notice the `coq.say` line in the tactic. The printed term is the following:
```prolog
app
 [(let `_` 
  (prod `x` (global (indt «nat»)) c0 \
      app [global (indt «eq»), X0 c0, c0, global (indc «O»)]) X1 c0 \ c0), 
  global (indc «O»)]
```
which is equivalent to (Coq syntax):
```coq
(let u : (forall x : nat, x = 0) := _ in u) 0
```

This term should be able to prove the first goal substitution (the one given by Coq, with `x = 0`), but elpi gives the one with `x = x`. Is it a bug or intentional behaviour?
Thanks.

### LINKED ISSUE #14972: Detyper and partial application
```coq
Set Printing All.
Check (plus : nat -> nat -> nat) O O.
Check (plus O : nat -> nat) O.
```
In the first case the cast is kept, in the latter it is lost.
I've reasons to believe it is the detyper that does this, but I'm not 100% sure.

CC @herbelin
See also: https://github.com/LPCIC/coq-elpi/issues/294

### PR DESCRIPTION
**Pull request by gares**
This PR makes `constr -> glob -> constr` a bit more reliable.

If this is is sufficiently CI-green that we can consider it.
If we get a full green CI, I'll try be even bolder and just remove the call, since I don't see what is the point of it (especially of its special case, which this PR makes less special). If there is a cast in the term, why would we want to erase it?

Fixes #14972
Fixes LPCIC/coq-elpi#294
Overlay https://github.com/mattam82/Coq-Equations/pull/433

### CODE REVIEW

**Comment by ppedrot** [comment]
I think that, apart for detyping, `collapse_appl` is not used a lot, let alone in contexts where casts might appear. So it's unlikely to break the CI. But this PR does look a bit like an XY problem. Why not fix directly the detyper rather than tweaking a primitive that is (unfortunately) used elsewhere?

**Comment by gares** [comment]
If this passes, then I'll remove the call from detyping, if this is what you ask.

**Comment by gares** [comment]
In general removing a cast from a term which is well typed should no affect its typability. I fixed only the case where the term under the cast is a hole (which detype, as I use it, turns it into a linear implicit.. I could keep the evar, but according to my experience evars withing glob are not working well).

File: engine/termops.ml [hunk 1]
```diff
@@ -980,9 +980,9 @@ let rec strip_outer_cast sigma c = match EConstr.kind sigma c with
 (* flattens application lists throwing casts in-between *)
 let collapse_appl sigma c = match EConstr.kind sigma c with
   | App (f,cl) ->
-    if EConstr.isCast sigma f then
+    if EConstr.isApp sigma f then
```
Comment (SkySkimmer) [inline engine/termops.ml:983]: Isn't this impossible?

Comment (gares) [reply engine/termops.ml:983]: What if f was an evar assigned to an App?

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@gares I've merged the upstream overlay so you can even rebase this to remove the overlay commit.

**Comment by ppedrot** [comment]
@coqbot merge now

**Pull request merged**